kernel/src/hil/digest.rs

```diff
@@ -56,13 +108,70 @@ pub trait Digest<'a, const L: usize> {
     /// error with error code ENOSUPPORT.
     fn run(&'a self, digest: &'static mut [u8; L])
         -> Result<(), (ErrorCode, &'static mut [u8; L])>;
+}
 
-    /// Clear the keys and any other sensitive data.
-    /// This won't clear the buffers provided to this API, that is up to the
-    /// user to clear.
-    fn clear_data(&self);
+/// Computes a digest (cryptographic hash) over data
+///
+/// 'L' is the length of the 'u8' array to store the digest output.
+pub trait DigestVerify<'a, const L: usize> {
+    /// Set the client instance which will receive the `verification_done()`
+    /// callback.
+    /// This is not required is using the `set_client()` fuction from the
+    /// `Digest` trait.
+    #[allow(unused_variables)]
+    fn set_verify_client(&'a self, client: &'a dyn ClientVerify<'a, L>) {}
+
+    /// Compare the specified digest in the `compare` buffer to the calculated
+    /// digest. This function is similar to `run()` and should be used instead
+    /// of `run()` if the caller doesn't need to know the output, just if it
+    /// matches a known value.
+    ///
+    /// For example:
+    /// ```rust,ignore
+    ///     // Compute a digest on data
+    ///     add_data(...);
+    ///     add_data(...);
+    ///
+    ///     // Compare the computed digest generated to an existing digest
+    ///     verify(...);
+    /// ```
+    /// NOTE: The above is just pseudo code. The user is expected to check for
+    /// errors and wait for the asyncronous calls to complete.
+    ///
+    /// The verify function is useful to compare input with a known digest
+    /// value. The verify function also saves callers from allocating a buffer
+    /// for the digest when they just need verification.
+    fn verify(
+        &'a self,
+        compare: &'static mut [u8; L],
+    ) -> Result<(), (ErrorCode, &'static mut [u8; L])>;
 }
 
+/// Computes a digest (cryptographic hash) over data
+///
+/// 'L' is the length of the 'u8' array to store the digest output.
+pub trait Digest<'a, const L: usize>:
+    DigestData<'a, L> + DigestHash<'a, L> + DigestVerify<'a, L>
+{
+    /// Set the client instance which will receive `hash_done()`,
+    /// `add_data_done()` and `verification_done()` callbacks.
+    fn set_client(&'a self, client: &'a dyn Client<'a, L>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+///
+/// 'L' is the length of the 'u8' array to store the digest output.
+pub trait DigestDataHash<'a, const L: usize>: DigestData<'a, L> + DigestHash<'a, L> {}
+
+impl<'a, T: DigestData<'a, L> + DigestHash<'a, L>, const L: usize> DigestDataHash<'a, L> for T {}
```

Comment:
Can a default implementation of `set_client` call the three underlying client set methods?

Reply:
Yes